# Web Scrapping Lab


## 1. Web scrapping using Beautiful Soup

Web scrapping is a practice in Data Science to extract data from webpages. In this lab, you will learn how to extract data in table forms from a webpage using Python's Beautiful Soup module.

Necessary libraries and modules to import are pandas, numpy, matplotlib, seaborn. To perform web scrapping, urllib.request library is used to open url. Beautiful soup package will extract data from an html file.

In [4]:
import pandas
import numpy
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

from urllib.request import urlopen
from bs4 import BeautifulSoup


In [5]:
url = "https://www.hubertiming.com/results/2017GPTR"
html = urlopen(url)

After getting the html of the web, the next step is to create an object in Python for Beautiful Soup from the html. This is done by passing the html to the BeautifulSoup() function. Then, the Beautiful Soup package will parse the html, which is taking the raw html text and break it into python objects.

In [6]:
soup = BeautifulSoup( html , 'lxml')
type(soup)

bs4.BeautifulSoup

Now we can extract information from the webpage such as title, text, html tags

In [7]:
# Get the title
title = soup.title
print(title)

# Get the text
text = soup.get_text()
#print(text)

<title>Race results for the 2017 Intel Great Place to Run \ Urban Clash Games!</title>


You can use the find_all() method of soup to extract useful html tags within a webpage. Examples of useful tags include < a > for hyperlinks, < table > for tables, < tr > for table rows, < th > for table headers, and < td > for table cells. The code below shows how to extract all the hyperlinks within the webpage. [1]

In [8]:
soup.find_all('a')

[<a href="mailto:timing@hubertiming.com">timing@hubertiming.com</a>,
 <a href="https://www.hubertiming.com">Huber Timing Home</a>,
 <a class="btn btn-primary btn-lg" href="/results/2017GPTR10K" role="button" style="margin: 0px 0px 5px 5px"><i aria-hidden="true" class="fa fa-user"></i> 10K</a>,
 <a class="btn btn-primary btn-lg" href="/results/summary/2017GPTR" role="button" style="margin: 0px 0px 5px 5px"><i class="fa fa-stream"></i> Summary</a>,
 <a id="individual" name="individual"></a>,
 <a data-url="/results/2017GPTR" href="#tabs-1" id="rootTab" style="font-size: 18px">5K Results</a>,
 <a href="https://www.hubertiming.com/"><img height="65" src="https://www.hubertiming.com//sites/all/themes/hubertiming/images/clockWithFinishSign_small.png" width="50"/>Huber Timing</a>,
 <a href="https://facebook.com/hubertiming/"><img src="https://www.hubertiming.com/results/FB-f-Logo__blue_50.png"/></a>]

From the output above, html tags sometimes have attributes like class, src, href, etc. These attributes provide additional information about html elements.

### Exercise 1

Use a loop and method get('href') to extract and print out only hyperlinks.

In [9]:
# Enter your code here

In [11]:
#Solution
# all_links = soup.find_all('a')
# for link in all_links:
#     print(link.get('href'))

Now, we will try print out table rows only. Remember from the examples of useful tags in find_all() method, < tr > stands for table rows. Pass the 'tr' attribute to soup.find_all().

In [12]:
rows = soup.find_all('tr')
print(rows[:10])

[<tr colspan="2">
<b>5K:</b>
</tr>, <tr>
<td>Finishers:</td>
<td>1458</td>
</tr>, <tr>
<td>Male:</td>
<td>771</td>
</tr>, <tr>
<td>Female:</td>
<td>687</td>
</tr>, <tr class="header">
<th>Place</th>
<th>Bib</th>
<th>Name</th>
<th>Gender</th>
<th>City</th>
<th>State</th>
<th>Time</th>
<th>Gun Time</th>
</tr>, <tr data-bib="2320">
<td>1</td>
<td>2320</td>
<td>

                    DANIEL M HINCKLEY

                </td>
<td>M</td>
<td>HILLSBORO</td>
<td>OR</td>
<td>16:42</td>
<td>16:44</td>
</tr>, <tr data-bib="2335">
<td>2</td>
<td>2335</td>
<td>

                    KORY F GRAY

                </td>
<td>M</td>
<td>HILLSBORO</td>
<td>OR</td>
<td>17:34</td>
<td>17:35</td>
</tr>, <tr data-bib="1770">
<td>3</td>
<td>1770</td>
<td>

                    FILIP SCHMOLE

                </td>
<td>M</td>
<td>PORTLAND</td>
<td>OR</td>
<td>18:13</td>
<td>18:14</td>
</tr>, <tr data-bib="2584">
<td>4</td>
<td>2584</td>
<td>

                    TRENTON C ROLLING

                </td>
<td>M</td>
<

The goal of this lab is to take a table from a webpage and convert it into a dataframe for easier manipulation using Python. To get there, you should get all table rows in list form first and then convert that list into a dataframe.

### Exercise 2

Use a loop and method find_all('td'), which is the html tag "sandwiching" our observation in each cell of the row, to create a list of table cells (td) of the rows.

In [13]:
# Enter your code here

In [16]:
# Solution
# for row in rows:
#     cell = row.find_all('td')
# print(cell)
# type(cell)

The output from your solution above shows that each row is printed with html tags embedded in each row. This is not what you want. You can use remove the html tags using Beautiful Soup or regualr expressions (regex).

The easiest way to remove html tags is to use Beautiful Soup, and it takes just one line of code to do it. Pass the string of interest into BeautifulSoup() and use the get_text() method to extract the text without html tags.

In [18]:
str_cells = str(cell)
clean_text = BeautifulSoup(str_cells, 'lxml').get_text()
print(clean_text)

[1458, 1400, 

                    SUMALATHA PURMA

                , F, PORTLAND, OR, 1:48:13, 1:48:13]


# 2. Regular Expression (Regex)

This method is also used to extract text without html tags, but it is highly discouraged because it's easy to make mistakes. Regular expression requires importing the re module. 

In [19]:
import re

list_rows = []
for row in rows:
    cells = row.find_all('td')                  #make a list of cells in row
    str_cells = str(cells)                      #convert them into string type
    clean = re.compile('<.*?>')                 #compile a regular expression by passing a string to match to re.compile. The '.*?' will match an opening angle bracket followed by anything and by a closing angle bracket.
    clean2 = (re.sub(clean, '',str_cells))      #after compile a regular expression, re.sub() method finds all the substrings where regular expression mathches and replace them with an empty string.
    list_rows.append(clean2)
print(clean2)
type(clean2)


[1458, 1400, 

                    SUMALATHA PURMA

                , F, PORTLAND, OR, 1:48:13, 1:48:13]


str

In [20]:
print(clean)

re.compile('<.*?>')


In [23]:
# print(list_rows)

Next, we will convert the list into a dataframe and get a quick view of the first 10 rows using Panda 

In [25]:
df = pandas.DataFrame(list_rows)
df.head(10)

0
0                                                 []
1                                 [Finishers:, 1458]
2                                       [Male:, 771]
3                                     [Female:, 687]
4                                                 []
5  [1, 2320, \r\n\r\n                    DANIEL M...
6  [2, 2335, \r\n\r\n                    KORY F G...
7  [3, 1770, \r\n\r\n                    FILIP SC...
8  [4, 2584, \r\n\r\n                    TRENTON ...
9  [5, 2688, \r\n\r\n                    YEAN-AN ...

## 3. Data Manipulation and Cleaning

The dataframe is not in the format we want. To clean it up, we will split the "0" column into multiple columns at the comma position, using str.split() method.

In [26]:
df1 = df[0].str.split(',', expand = True)
df1.head(10)

0       1                                                  2  \
0           []    None                                               None   
1  [Finishers:   1458]                                               None   
2       [Male:    771]                                               None   
3     [Female:    687]                                               None   
4           []    None                                               None   
5           [1    2320   \r\n\r\n                    DANIEL M HINCKLEY...   
6           [2    2335   \r\n\r\n                    KORY F GRAY\r\n\r...   
7           [3    1770   \r\n\r\n                    FILIP SCHMOLE\r\n...   
8           [4    2584   \r\n\r\n                    TRENTON C ROLLING...   
9           [5    2688   \r\n\r\n                    YEAN-AN LIAO\r\n\...   

      3           4     5       6        7  
0  None        None  None    None     None  
1  None        None  None    None     None  
2  None        None  None    None     None  
3  None        None  None    None     None  
4  None        None  None    None     None  
5     M   HILLSBORO    OR   16:42   16:44]  
6     M   HILLSBORO    OR   17:34   17:35]  
7     M    PORTLAND    OR   18:13   18:14]  
8     M    PORTLAND    OR   18:32   18:35]  
9     M   HILLSBORO    OR   19:12   19:18]

The table is missing headers. You can get all table headers by passing argument 'th' into find_all() method. Then, get rid of the html tag by doing similarly to table rows.

In [29]:
table_headers = soup.find_all('th')
print(table_headers)

[<th>Place</th>, <th>Bib</th>, <th>Name</th>, <th>Gender</th>, <th>City</th>, <th>State</th>, <th>Time</th>, <th>Gun Time</th>]


In [30]:
headers = []
str_header = str(table_headers)
cleantext2 = BeautifulSoup(str_header, 'lxml').get_text()
headers.append(cleantext2)
print(headers)

['[Place, Bib, Name, Gender, City, State, Time, Gun Time]']


You can then convert the list of headers into a pandas dataframe.

In [31]:
df2 = pandas.DataFrame(headers)
df2.head()

0
0  [Place, Bib, Name, Gender, City, State, Time, ...

In [32]:
df3 = df2[0].str.split(',', expand = True)
df3.head()

0     1      2        3      4       5      6           7
0  [Place   Bib   Name   Gender   City   State   Time   Gun Time]

The 2 dataframes can be converted into one using the concat() method.

In [33]:
frames = [df3, df1]
df4 = pandas.concat(frames)
df4.head(10)

0       1                                                  2  \
0       [Place     Bib                                               Name   
0           []    None                                               None   
1  [Finishers:   1458]                                               None   
2       [Male:    771]                                               None   
3     [Female:    687]                                               None   
4           []    None                                               None   
5           [1    2320   \r\n\r\n                    DANIEL M HINCKLEY...   
6           [2    2335   \r\n\r\n                    KORY F GRAY\r\n\r...   
7           [3    1770   \r\n\r\n                    FILIP SCHMOLE\r\n...   
8           [4    2584   \r\n\r\n                    TRENTON C ROLLING...   

         3           4       5       6           7  
0   Gender        City   State    Time   Gun Time]  
0     None        None    None    None        None  
1     None        None    None    None        None  
2     None        None    None    None        None  
3     None        None    None    None        None  
4     None        None    None    None        None  
5        M   HILLSBORO      OR   16:42      16:44]  
6        M   HILLSBORO      OR   17:34      17:35]  
7        M    PORTLAND      OR   18:13      18:14]  
8        M    PORTLAND      OR   18:32      18:35]

Assign the first row to be the table header.

In [35]:
df5 = df4.rename(columns=df4.iloc[0])
df5.head()

[Place     Bib   Name   Gender   City   State   Time   Gun Time]
0       [Place     Bib   Name   Gender   City   State   Time   Gun Time]
0           []    None   None     None   None    None   None        None
1  [Finishers:   1458]   None     None   None    None   None        None
2       [Male:    771]   None     None   None    None   None        None
3     [Female:    687]   None     None   None    None   None        None

At this point, the table is almost preperly formatted. For analysis, you can start by getting an overview of the data as shown below.

In [36]:
df5.info()
df5.shape

<class 'pandas.core.frame.DataFrame'>
Index: 1464 entries, 0 to 1462
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   [Place      1464 non-null   object
 1    Bib        1462 non-null   object
 2    Name       1459 non-null   object
 3    Gender     1459 non-null   object
 4    City       1459 non-null   object
 5    State      1459 non-null   object
 6    Time       1459 non-null   object
 7    Gun Time]  1459 non-null   object
dtypes: object(8)
memory usage: 102.9+ KB


(1464, 8)

The table has 1464 rows and 8 columns. You can drop all rows with missing values by using dropna().

In [37]:
df6 = df5.dropna(axis=0, how='any')

In [38]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1464 entries, 0 to 1462
Data columns (total 8 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   [Place      1464 non-null   object
 1    Bib        1462 non-null   object
 2    Name       1459 non-null   object
 3    Gender     1459 non-null   object
 4    City       1459 non-null   object
 5    State      1459 non-null   object
 6    Time       1459 non-null   object
 7    Gun Time]  1459 non-null   object
dtypes: object(8)
memory usage: 102.9+ KB


Now we will remove the duplicate headers by also using drop().

In [39]:
df7 = df6.drop(df6.index[0])
df7.head()

[Place    Bib                                               Name  Gender  \
5     [1   2320   \r\n\r\n                    DANIEL M HINCKLEY...       M   
6     [2   2335   \r\n\r\n                    KORY F GRAY\r\n\r...       M   
7     [3   1770   \r\n\r\n                    FILIP SCHMOLE\r\n...       M   
8     [4   2584   \r\n\r\n                    TRENTON C ROLLING...       M   
9     [5   2688   \r\n\r\n                    YEAN-AN LIAO\r\n\...       M   

         City  State    Time  Gun Time]  
5   HILLSBORO     OR   16:42     16:44]  
6   HILLSBORO     OR   17:34     17:35]  
7    PORTLAND     OR   18:13     18:14]  
8    PORTLAND     OR   18:32     18:35]  
9   HILLSBORO     OR   19:12     19:18]

The final step is to clean the '[' and ']' in column headers Place and Time, as well as in their columns cells using rename() and str.strip() methods.

In [54]:
df7.rename(columns={'[Place': 'Place'}, inplace = True)
df7.rename(columns={' Gun Time]': 'Gun Time'}, inplace = True)  #make sure to include a space before the quotation mark in 'Team]'.
df7.head()


Place    Bib                                               Name  Gender  \
5     1   2320   \r\n\r\n                    DANIEL M HINCKLEY...       M   
6     2   2335   \r\n\r\n                    KORY F GRAY\r\n\r...       M   
7     3   1770   \r\n\r\n                    FILIP SCHMOLE\r\n...       M   
8     4   2584   \r\n\r\n                    TRENTON C ROLLING...       M   
9     5   2688   \r\n\r\n                    YEAN-AN LIAO\r\n\...       M   

         City  State    Time Gun Time  
5   HILLSBORO     OR   16:42    16:44  
6   HILLSBORO     OR   17:34    17:35  
7    PORTLAND     OR   18:13    18:14  
8    PORTLAND     OR   18:32    18:35  
9   HILLSBORO     OR   19:12    19:18

In [55]:
df7['Gun Time'] = df7['Gun Time'].str.strip(']')
df7.head()

Place    Bib                                               Name  Gender  \
5     1   2320   \r\n\r\n                    DANIEL M HINCKLEY...       M   
6     2   2335   \r\n\r\n                    KORY F GRAY\r\n\r...       M   
7     3   1770   \r\n\r\n                    FILIP SCHMOLE\r\n...       M   
8     4   2584   \r\n\r\n                    TRENTON C ROLLING...       M   
9     5   2688   \r\n\r\n                    YEAN-AN LIAO\r\n\...       M   

         City  State    Time Gun Time  
5   HILLSBORO     OR   16:42    16:44  
6   HILLSBORO     OR   17:34    17:35  
7    PORTLAND     OR   18:13    18:14  
8    PORTLAND     OR   18:32    18:35  
9   HILLSBORO     OR   19:12    19:18